src/Generics/Linear/TH.hs

```diff
@@ -0,0 +1,539 @@
+-- Alternatively, one could type $(deriveGeneric1 'FamilyFalse)
+@
+-}
+
+-- Adapted from Generics.Regular.TH, via
+-- Generics.Deriving.TH
+module Generics.Linear.TH (
+      deriveGeneric
+    , deriveGeneric1
+    , deriveGenericAnd1
+  ) where
+
+import           Control.Monad ((>=>), unless, when)
+
+import qualified Data.Map as Map
+
+import           Generics.Linear.TH.Internal
+import           Generics.Linear.TH.MetaData
+import           Language.Haskell.TH.Datatype
+import           Language.Haskell.TH.Lib
+import           Language.Haskell.TH
+
+-- Imports for splices
+import           Generics.Linear.Class
+  hiding ( uAddr#, uChar#, uDouble#, uFloat#, uInt#, uWord#
+         , unM1, unK1, unPar1, unComp1)
+import           Generics.Linear.TH.Insertions
+  hiding ((.))
+import qualified Generics.Linear.TH.Insertions as Ins
+import           GHC.Exts (Addr#, Char#, Int#, Word#, Double#, Float#)
+
+-- | Given the name of a type or data family constructor,
+-- derive a 'Generic' instance.
+deriveGeneric :: Name -> Q [Dec]
+deriveGeneric = deriveGenericCommon True False
+
+-- | Given the name of a type or data family constructor,
+-- derive a 'Generic1' instance.
+deriveGeneric1 :: Name -> Q [Dec]
+deriveGeneric1 = deriveGenericCommon False True
+
+-- | Given the name of a type or data family constructor,
+-- derive a 'Generic' instance and a 'Generic1' instance.
+deriveGenericAnd1 :: Name -> Q [Dec]
+deriveGenericAnd1 = deriveGenericCommon True True
+
+deriveGenericCommon :: Bool -> Bool -> Name -> Q [Dec]
+deriveGenericCommon generic generic1 n = do
+    b <- if generic
+            then deriveInst Generic n
+            else return []
+    c <- if generic1
+            then deriveInst Generic1 n
+            else return []
+    return (b ++ c)
+
+deriveInst :: GenericClass -> Name -> Q [Dec]
+deriveInst Generic  = deriveInstCommon ''Generic  ''Rep  Generic  'from  'to
+deriveInst Generic1 = deriveInstCommon ''Generic1 ''Rep1 Generic1 'from1 'to1
+
+deriveInstCommon :: Name
+                 -> Name
+                 -> GenericClass
+                 -> Name
+                 -> Name
+                 -> Name
+                 -> Q [Dec]
+deriveInstCommon genericName repName gClass fromName toName n = do
+  i <- reifyDataInfo n
+  let (name, instTys, cons, dv) = either error id i
+  (origTy, origKind) <- buildTypeInstance gClass name instTys
+  tyInsRHS <- makeRepInline   gClass dv name instTys cons origTy
+
+  let origSigTy = SigT origTy origKind
+  tyIns <- tySynInstDCompat repName Nothing [return origSigTy] (return tyInsRHS)
+  let
+    mkBody maker = [clause [] (normalB $
+      lamCaseE [maker gClass 1 1 instTys cons]) []]
+
+    fcs = mkBody mkFrom
+    tcs = mkBody mkTo
+
+  fmap (:[]) $
+    instanceD (cxt []) (conT genericName `appT` return origSigTy)
+                         [return tyIns, funD fromName fcs, funD toName tcs]
+
+makeRepInline :: GenericClass
```

Comment:
For what it's worth, I discovered recently while working on dreixel/generic-deriving#80 that you don't actually need this `makeRepInline` function if you only use the `derive-` functions. I'm hoping to get up a patch soon. Removing `makeRepInline` would allow removing a fair bit of needless complexity in some of the downstream functions—for instance, you wouldn't need to propagate a `TypeSubst` down to `repField`. anymore.